src/compiler/scala/tools/nsc/typechecker/Implicits.scala

```diff
@@ -1388,27 +1383,30 @@ trait Implicits {
         }
       }
       if (result.isSuccess && isView) {
-        def maybeInvalidConversionError(msg: String) {
+        def maybeInvalidConversionError(msg: String): Boolean = {
           // We have to check context.ambiguousErrors even though we are calling "issueAmbiguousError"
           // which ostensibly does exactly that before issuing the error. Why? I have no idea. Test is pos/t7690.
           // AM: I would guess it's because ambiguous errors will be buffered in silent mode if they are not reported
           if (context.ambiguousErrors)
             context.issueAmbiguousError(AmbiguousImplicitTypeError(tree, msg))
+          true
         }
         pt match {
-          case Function1(_, out) =>
-            // must inline to avoid capturing result
-            def prohibit(sym: Symbol) = (sym.tpe <:< out) && {
-              maybeInvalidConversionError(s"the result type of an implicit conversion must be more specific than ${sym.name}")
-              true
-            }
-            if (prohibit(AnyRefClass) || (settings.isScala211 && prohibit(AnyValClass)))
-              result = SearchFailure
-          case _                 => false
-        }
-        if (settings.isScala211 && isInvalidConversionSource(pt)) {
-          maybeInvalidConversionError("an expression of type Null is ineligible for implicit conversion")
-          result = SearchFailure
+          // SI-10206 don't use subtyping to rule out AnyRef/AnyVal: there are several valid structural types that are supertypes of AnyRef (e.g., created by HasMember);
+          // typeSymbol will do the trick (AnyRef is a type alias for Object)
+          case Function1(in, out)  =>
+            val outSym = out.typeSymbol
```

Comment:
I think I like this approach better than mine. Another thing I was thinking about in was whether we might accidentally be contributing unwanted constraints to type variables in `out` in the old approach using `<:<`. Your approach also avoids that.

Noting that if `out` is itself a refinement type of `AnyRef`, `.typeSymbol` will not give us `AnyRefClass` (which is a good thing here, as we don't want to disallow that.

One corner case might be `AnnotatedTypes` with an annotation checker that treats `AnyRef @foo`  as distinct from `AnyRef` and uses implicits to model conversions between the two?